src/registry/PartialRegistry.ts

```diff
@@ -0,0 +1,74 @@
+import type { Logger } from 'pino';
+
+import type { ChainMap, ChainMetadata, ChainName, WarpCoreConfig } from '@hyperlane-xyz/sdk';
+import { ChainAddresses } from '../types.js';
+import { ChainFiles, IRegistry, RegistryContent, RegistryType } from './IRegistry.js';
+import { SynchronousRegistry } from './SynchronousRegistry.js';
+
+const PARTIAL_URI_PLACEHOLDER = '__partial_registry__';
+
+/**
+ * A registry that accepts partial data, such as incomplete chain metadata or addresses.
+ * Useful for merging with other registries force overrides of subsets of data.
+ */
+export interface PartialRegistryOptions {
+  chainMetadata?: ChainMap<Partial<ChainMetadata>>;
+  chainAddresses?: ChainMap<Partial<ChainAddresses>>;
+  // TODO add more fields here as needed
+  logger?: Logger;
+}
+
+export class PartialRegistry extends SynchronousRegistry implements IRegistry {
+  public readonly type = RegistryType.Partial;
+  public chainMetadata: ChainMap<Partial<ChainMetadata>>;
+  public chainAddresses: ChainMap<Partial<ChainAddresses>>;
+
+  constructor({ chainMetadata, chainAddresses, logger }: PartialRegistryOptions) {
+    super({ uri: PARTIAL_URI_PLACEHOLDER, logger });
+    this.chainMetadata = chainMetadata || {};
+    this.chainAddresses = chainAddresses || {};
+  }
+
+  listRegistryContent(): RegistryContent {
+    const chains: ChainMap<ChainFiles> = {};
+    Object.keys(this.chainMetadata).forEach((c) => {
+      chains[c] ||= {};
+      chains[c].metadata = PARTIAL_URI_PLACEHOLDER;
+    });
+    Object.keys(this.chainAddresses).forEach((c) => {
+      chains[c] ||= {};
+      chains[c].addresses = PARTIAL_URI_PLACEHOLDER;
+    });
+    return {
+      chains,
+      deployments: {},
+    };
+  }
+
+  getMetadata(): ChainMap<ChainMetadata> {
+    return this.chainMetadata as ChainMap<ChainMetadata>;
```

Comment:
I assume it's intentional to accept that it'd be possible for someone to end up with a PartialRegistry treated as if it were a full registry, & that dealing with fully accurate types wasn't worth the hassle?

Reply:
Exactly. It's easy to use this wrong but the I don't want to weaken the types everywhere else just to accommodate this thing that we'll probably be the only ones to use